### PR TITLE
Get num particles

### LINKED ISSUE #37: deletion of particles does not set VerletLists::_neighborListIsValid to false
**Describe the bug**
deletion of particles does not set VerletLists::_neighborListIsValid to false

**To Reproduce**

**Expected behavior**
deletion of particles should set VerletLists::_neighborListIsValid to false

**Problems**
a) thread safety
b) the iterators don't know their parent container

**Additional context**
ref #10

### LINKED ISSUE #216: AutoPas::getNumberOfParticles() returns number of saved particles
**Describe the bug**
`AutoPas::getNumberOfParticles()` currently returns the total (owned + halo + (dummy)) amount of particles within a container instead of the number particles owned by the container. 

**To Reproduce**
1. Add particles + halo particles to AutoPas. 
2. `getNumberOfParticles()` should then return the number of owned particles

**Expected behavior**
`getNumberOfParticles()` should return the number of owned particles.

**Additional context**
A test should be added for this!

### PR DESCRIPTION
**Pull request by FG-TUM**
# Description

- [x] add tracker for the number of owned and halo particles each.
- [x] getter from AutoPas interface
- [x] iteratio.deleteCurrentParticle will not be caught -> AutoPas.deleteParticle(iter)
- [x] iteratio.deleteCurrentParticle "hidden" in internal namespace.

## Resolved Issues

- fixes #216 
- fixes #37 for real

# How Has This Been Tested?

- [x] new AutoPasTest.getNumParticlesTest

### CODE REVIEW

File: tests/testAutopas/tests/autopasInterface/AutoPasTest.cpp [hunk 1]
```diff
@@ -191,4 +191,32 @@ TEST_F(AutoPasTest, checkConstIterator) {
...
+TEST_F(AutoPasTest, getNumParticlesTest) {
+  expectedParticles(0, 0);
+
+  Particle particle;
+
+  particle.setR({1, 1, 1});
+  autoPas.addParticle(particle);
+  expectedParticles(1, 0);
+
+  particle.setR({-0.1, -0.1, -0.1});
```
Comment (SteffenSeckler) [inline tests/testAutopas/tests/autopasInterface/AutoPasTest.cpp]:
```suggestion
  // add second particle
  particle.setR({-0.1, -0.1, -0.1});
```

File: tests/testAutopas/tests/autopasInterface/AutoPasTest.cpp [hunk 2]
```diff
@@ -191,4 +191,32 @@ TEST_F(AutoPasTest, checkConstIterator) {
...
+  EXPECT_EQ(autoPas.getNumberOfParticles(autopas::IteratorBehavior::ownedOnly), expectedOwned);
+  EXPECT_EQ(autoPas.getNumberOfParticles(autopas::IteratorBehavior::haloOnly), expectedHalo);
+}
+
+TEST_F(AutoPasTest, getNumParticlesTest) {
+  expectedParticles(0, 0);
+
+  Particle particle;
+
+  particle.setR({1, 1, 1});
```
Comment (SteffenSeckler) [inline tests/testAutopas/tests/autopasInterface/AutoPasTest.cpp]:
```suggestion
  // add first particle
  particle.setR({1, 1, 1});
```

File: tests/testAutopas/tests/autopasInterface/AutoPasTest.cpp [hunk 3]
```diff
@@ -191,4 +191,32 @@ TEST_F(AutoPasTest, checkConstIterator) {
...
+  autoPas.addParticle(particle);
+  expectedParticles(1, 0);
+
+  particle.setR({-0.1, -0.1, -0.1});
+  autoPas.addOrUpdateHaloParticle(particle);
+  expectedParticles(1, 1);
+
+  // update container is expected to remove all halo particles
+  auto unused = autoPas.updateContainerForced();
+  expectedParticles(1, 0);
```
Comment (SteffenSeckler) [inline tests/testAutopas/tests/autopasInterface/AutoPasTest.cpp]:
```suggestion
  expectedParticles(1, 0);
  EXPECT_EQUALS(0, unused.size());
```

File: tests/testAutopas/tests/autopasInterface/AutoPasTest.cpp [hunk 4]
```diff
@@ -191,4 +191,32 @@ TEST_F(AutoPasTest, checkConstIterator) {
...
+
+  autoPas.begin()->setR({-0.2, -0.2, -0.2});
+  unused = autoPas.updateContainerForced();
```
Comment (SteffenSeckler) [inline tests/testAutopas/tests/autopasInterface/AutoPasTest.cpp]:
```suggestion
  unused = autoPas.updateContainerForced();
  EXPECT_EQUALS(1, unused.size());
```

File: src/autopas/LogicHandler.h [hunk 5]
```diff
@@ -109,6 +114,8 @@ class LogicHandler {
   void deleteAllParticles() {
     _containerIsValid = false;
     _autoTuner.getContainer()->deleteAllParticles();
+    _numParticlesOwned.exchange(0, std::memory_order_relaxed);
```
Comment (SteffenSeckler) [inline src/autopas/LogicHandler.h]: comment why!

File: src/autopas/LogicHandler.h [hunk 6]
```diff
@@ -38,7 +38,10 @@ class LogicHandler {
     if (not isContainerValid() or forced) {
       AutoPasLog(debug, "Initiating container update.");
       _containerIsValid = false;
-      return std::make_pair(std::move(_autoTuner.getContainer()->updateContainer()), true);
+      _numParticlesHalo.exchange(0, std::memory_order_relaxed);
+      auto returnPair = std::make_pair(std::move(_autoTuner.getContainer()->updateContainer()), true);
+      _numParticlesOwned.fetch_sub(returnPair.first.size(), std::memory_order_relaxed);
```
Comment (SteffenSeckler) [inline src/autopas/LogicHandler.h]: comment

**Comment by SteffenSeckler** [comment]
the variant with passing the iterator to the AutoPas object would be sort of similar to:
https://en.wikipedia.org/wiki/Erase%E2%80%93remove_idiom

File: src/autopas/AutoPas.h [hunk 7]
```diff
@@ -171,6 +171,12 @@ class AutoPas {
    */
   void deleteAllParticles() { _logicHandler->deleteAllParticles(); }
 
+  /**
+   * Deletes the particle behind the current iterator position.
```
Comment (SteffenSeckler) [inline src/autopas/AutoPas.h]: comment that this only works if it isn't a const iterator

File: src/autopas/iterators/ParticleIteratorInterface.h [hunk 8]
```diff
@@ -57,6 +57,7 @@ class ParticleIteratorInterface {
 
   /**
    * Deletes the current particle.
+   * @note Do NOT use this function from outside the autopas namespce. Use AutoPas::deleteParticle instead.
```
[the remark this answers is not in the document]
Comment (SteffenSeckler) [reply src/autopas/iterators/ParticleIteratorInterface.h]: adapt comment

File: src/autopas/LogicHandler.h [hunk 9]
```diff
@@ -109,6 +118,22 @@ class LogicHandler {
...
+    // all particles are gone -> reset counters.
+    _numParticlesOwned.exchange(0, std::memory_order_relaxed);
+    _numParticlesHalo.exchange(0, std::memory_order_relaxed);
+  }
+
+  /**
+   * Deletes a single particle and updates internal particle counters.
+   * @param iter
+   */
+  void deleteParticle(ParticleIteratorWrapper<Particle, true> &iter) {
```
[the remark this answers is not in the document]
Comment (FG-TUM) [reply src/autopas/LogicHandler.h]: will be invalidated

**Pull request merged**